File: neet_code_150/src/two_pointers/trapping_rain_water.rs
// 解けなかった
struct Solution {}
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        1
    }

    pub fn trap_sq(height: Vec<i32>) -> i32 {
        let n = height.len();
        let mut result = 0;

        for i in 0..n {}

        result
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn trap(height: Vec<i32>) -> i32 {
        if height.is_empty() {
            return 0;
        }
        let (mut left, mut right) = (0, height.len() - 1);
        let (mut left_max, mut right_max) = (height[left], height[right]);
        let mut result = 0;

        while left < right {
            if left_max < right_max {
                left += 1;
                left_max = left_max.max(height[left]);
                result += left_max - height[left];
            } else {
                right -= 1;
                right_max = right_max.max(height[right]);
                result += right_max - height[right];
            }
        }
        result
    }
}

// space: O(n), time: O(n) latter
struct SolutionLatter;
impl SolutionLatter {
    // AC
    pub fn trap(height: Vec<i32>) -> i32 {
        let n = height.len();
        if height.is_empty() {
            return 0;
        }

        let mut max_left = vec![i32::MIN; n + 1];
        let mut max_right = vec![i32::MIN; n + 1];
        for i in 0..n {
            max_left[i + 1] = max_left[i].max(height[i]);
            max_right[n - 1 - i] = max_right[n - i].max(height[n - 1 - i]);
        }

        /*
        println!("max_left: {:?}", max_left);
        println!("max_right: {:?}", max_right);
        */
        let mut result = 0;
        for i in 0..n {
            result += (max_left[i + 1].min(max_right[i]) - height[i]).max(0);
        }

        result
    }
}

fn main() {
    let case_1 = vec![0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1];
    // => 6
    let case_2 = vec![4, 2, 0, 3, 2, 5];
    // => 9
    let case_3 = vec![0, 1, 0, 0, 1, 0, 1];
    // => 3

    println!("case_1: {}", SolutionAns::trap(case_1.clone()));
    println!("case_2: {}", SolutionAns::trap(case_2.clone()));
    println!("case_3: {}", SolutionAns::trap(case_3.clone()));

    println!("case_1: {}", SolutionLatter::trap(case_1.clone()));
    println!("case_2: {}", SolutionLatter::trap(case_2.clone()));
    println!("case_3: {}", SolutionLatter::trap(case_3.clone()));

    println!("case_1: {}", SolutionLatter::trap_2(case_1.clone()));
    /*
    println!("case_2: {}", SolutionLatter::trap_2(case_2.clone()));
    println!("case_3: {}", SolutionLatter::trap_2(case_3.clone()));
    */
}

File: neet_code_150_cpp/two_pointers/trapping_rain_water.cpp
#include <iostream>
#include <vector>

// 模範解答
class SolutionAns {
    public:
        int trap(std::vector<int>& height) {
            auto [i,j] = std::pair(0, height.size() - 1);
            auto [maxLeft, maxRight] = std::pair(height[i], height[j]);

            int result = 0;
            while(i < j) {
                if(maxLeft <= maxRight) {
                    ++i;
                    maxLeft = std::max(maxLeft, height[i]);
                    result += maxLeft - height[i];
                } else {
                    --j;
                    maxRight = std::max(maxRight, height[j]);
                    result += maxRight - height[j];
                }
            }

            return result;
        }
};

// 後で解いた解
// AC
class SolutionLatter {
    public:
        int trap(std::vector<int>& height) {
            int result = 0;
            int left = 0, right = height.size() - 1;
            int left_max = 0, right_max = 0;
            while(left < right) {
                if(height[left] < height[right]) {
                    left_max = std::max(left_max, height[left]);
                    result += left_max - height[left];
                    ++left;
                } else if(height[left] >= height[right]) {
                    right_max = std::max(right_max, height[right]);
                    result += right_max - height[right];
                    --right;
                }
            }

            return result;
        }
};

int main(void) {
    std::vector<int> case_1 = {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
    // => 6
    std::vector<int> case_2 = {4, 2, 0, 3, 2, 5};
    // => 9
    std::vector<int> case_3 = {0, 1, 0, 0, 1, 0, 1};
    // => 3

    SolutionAns s_ans;
    std::cout << s_ans.trap(case_1) << std::endl;
    std::cout << s_ans.trap(case_2) << std::endl;
    std::cout << s_ans.trap(case_3) << std::endl;

    SolutionLatter s_ans_latter;
    std::cout << s_ans_latter.trap(case_1) << std::endl;
    std::cout << s_ans_latter.trap(case_2) << std::endl;
    std::cout << s_ans_latter.trap(case_3) << std::endl;
}
